### PR TITLE
filter row_id with -1

### PR DESCRIPTION
**Pull request by julesbarbosa**
## What does this PR do?

when a chunk is empty, filter out the row with -1

## Changes

-   [x] Filter row_id

## Before reviewing

-   [x] Are there any linked PRs?
-   [ ] Did you test this PR?
-   [ ] Is there any migration or seeding necessary?
-   [x] Did you add api to only one of the language (python/typescript)?

### Links:

-   https://myaccern.atlassian.net/browse/TAD-2969
-   https://github.com/Accern/xyme-backend/pull/142

### New API added that not implemented in other languages

If your answer is yes, please also update the todo_api.txt at where the API is missing.

### CODE REVIEW

File: packages/python/accern_xyme/accern_xyme.py [hunk 1]
```diff
@@ -2142,7 +2156,13 @@ def get_content(self) -> Optional[ByteResponse]:
                         "uri": self.get_uri(),
                     })
                 self._ctype = ctype
-                return interpret_ctype(fin, ctype)
+                content = interpret_ctype(fin, ctype)
+                if filter_id and isinstance(content, pd.DataFrame):
+                    res = self.get_parent()
+                    row_id = res.get_info()
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2162]: info?

Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:2162]: also use and apply the index from the info

Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:2162]: like what we do in the backend

Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:2162]: move to read

File: packages/python/accern_xyme/accern_xyme.py [hunk 2]
```diff
@@ -2110,6 +2113,17 @@ def get_path(self, *path: str) -> 'BlobHandle':
         return BlobHandle(
             self._client, f"{self._uri}/{'/'.join(path)}", is_full=True)
 
+    def get_parent(self) -> 'BlobHandle':
+        if self._parent is None:
+            parsed = parse.urlparse(self._uri)
+            path = "/".join(parsed.path.split("/")[1:3])
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2119]: how can you do `1:3`? what is 0?

File: packages/python/accern_xyme/accern_xyme.py [hunk 3]
```diff
@@ -2110,6 +2113,17 @@ def get_path(self, *path: str) -> 'BlobHandle':
...
+            parsed = parsed._replace(path=path)
+            uri = parsed.geturl()
+            res = BlobHandle(
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2122]: one line?

File: packages/python/accern_xyme/accern_xyme.py (hunk 2 above)
Comment (julesbarbosa) [reply packages/python/accern_xyme/accern_xyme.py:2123]: the 0 is the "/" 
when I do the `geturl()` they add the "/"
but I'll return the 0 just to make sure.

File: packages/python/accern_xyme/accern_xyme.py [hunk 4]
```diff
@@ -1895,8 +1896,12 @@ def read(
             self,
             key: str,
             chunk: Optional[int],
-            force_refresh: bool = False) -> Optional[ByteResponse]:
-        return self.read_blob(key, chunk, force_refresh).get_content()
+            force_refresh: bool = False,
+            filter_id: bool = True) -> Optional[ByteResponse]:
+        content = self.read_blob(key, chunk, force_refresh).get_content()
+        if filter_id and isinstance(content, pd.DataFrame):
+            content = content[~content["row_id"] < 0]
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:1903]: also add set_index -- like in the backend.
also why ~ and < not >=?

Comment (julesbarbosa) [reply packages/python/accern_xyme/accern_xyme.py:1903]: set index with which column?

File: packages/python/accern_xyme/accern_xyme.py [hunk 5]
```diff
@@ -2138,6 +2144,16 @@ def get_path(self, *path: str) -> 'BlobHandle':
         return BlobHandle(
             self._client, f"{self._uri}/{'/'.join(path)}", is_full=True)
 
+    def get_parent(self) -> 'BlobHandle':
+        if self._parent is None:
```
Comment (jckhang) [inline packages/python/accern_xyme/accern_xyme.py:2148]: image
is this what you are trying to do?

Comment (jckhang) [reply packages/python/accern_xyme/accern_xyme.py:2148]: also add get_parent to todo_api for typescript

Comment (julesbarbosa) [reply packages/python/accern_xyme/accern_xyme.py:2148]: yes

File: packages/python/accern_xyme/accern_xyme.py [hunk 6]
```diff
@@ -1895,8 +1896,12 @@ def read(
             self,
             key: str,
             chunk: Optional[int],
-            force_refresh: bool = False) -> Optional[ByteResponse]:
-        return self.read_blob(key, chunk, force_refresh).get_content()
+            force_refresh: bool = False,
+            filter_id: bool = True) -> Optional[ByteResponse]:
+        content = self.read_blob(key, chunk, force_refresh).get_content()
+        if filter_id and isinstance(content, pd.DataFrame):
+            content = content[content["row_id"] >= 0]
```
Comment (julesbarbosa) [inline packages/python/accern_xyme/accern_xyme.py:1903]: set_index("index")

Comment (julesbarbosa) [reply packages/python/accern_xyme/accern_xyme.py:1903]: same as backend

File: packages/python/accern_xyme/accern_xyme.py [hunk 7]
```diff
@@ -2136,6 +2144,14 @@ def get_path(self, *path: str) -> 'BlobHandle':
         return BlobHandle(
             self._client, f"{self._uri}/{'/'.join(path)}", is_full=True)
 
+    def get_parent(self) -> 'BlobHandle':
+        if self._parent is None:
+            path = PurePath(self._uri)
+            uri = str(path.parent)
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2150]: `s3://accern-ml-titan@aws/dag/b8321c98d96894bc7b77d31ea29efc016/foo/bar/baz.json` -> full URI
parent gives us a non-full URI
`s3://accern-ml-titan@aws/dag/b8321c98d96894bc7b77d31ea29efc016`
path[0:2]

Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2150]:
```suggestion
            uri = "/".join(path.parts[:2])
```

File: packages/python/accern_xyme/accern_xyme.py (hunk 2 above)
Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py]:
```
>>> "/".join(pathlib.PurePath(parsed.path).parts[1:3])
```

File: packages/python/accern_xyme/accern_xyme.py [hunk 8]
```diff
@@ -2136,6 +2144,14 @@ def get_path(self, *path: str) -> 'BlobHandle':
         return BlobHandle(
             self._client, f"{self._uri}/{'/'.join(path)}", is_full=True)
 
+    def get_parent(self) -> 'BlobHandle':
+        if self._parent is None:
+            path = PurePath(self._uri)
+            uri = "/".join(path.parts[:2])
+            res = BlobHandle(self._client, uri, is_full=False)
+            self._parent = res
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2152]:
```
from urllib.parse import urlparse, urlunparse
from pathlib import PosixPath
s3_url = urlparse("s3://accern-ml-titan@aws/dag/b8321c98d96894bc7b77d31ea29efc016/foo/bar/baz.json")
s3_path = PurePath(*PosixPath(s3_url.path).parts[:3])
s3_new_url = urlunparse((s3_url.scheme, s3_url.netloc, s3_path.as_posix(), None, None, None))
```

**Review by JosuaKrause — APPROVED** [review]
lgtm

**Pull request merged**